src/main/java/org/mockito/internal/configuration/injection/filter/TypeBasedCandidateFilter.java

```diff
@@ -17,6 +25,52 @@ public TypeBasedCandidateFilter(MockCandidateFilter next) {
         this.next = next;
     }
 
+    protected boolean isCompatibleTypes(Type typeToMock, Type mockType) {
+        if (typeToMock instanceof ParameterizedType && mockType instanceof ParameterizedType) {
+            // ParameterizedType.equals() is documented as:
+            // "Instances of classes that implement this interface must implement
+            // an equals() method that equates any two instances that share the
+            // same generic type declaration and have equal type parameters."
+            // Unfortunately, e.g. Wildcard parameter "?" doesn't equal java.lang.String
+            if (typeToMock.equals(mockType)) {
+                return true;
+            } else {
+                ParameterizedType genericTypeToMock = (ParameterizedType) typeToMock;
+                ParameterizedType genericMockType = (ParameterizedType) mockType;
+                Type[] actualTypeArguments = genericTypeToMock.getActualTypeArguments();
+                Type[] actualTypeArguments2 = genericMockType.getActualTypeArguments();
+                // getRawType() says "the Type object representing the class or interface that
+                // declares this type",
+                // no clue why that's a Type rather than a Class as return type anyway
+                Class rawType = (Class) genericTypeToMock.getRawType();
+                Class rawType2 = (Class) genericMockType.getRawType();
+                if (Objects.equals(genericTypeToMock.getOwnerType(), genericMockType.getOwnerType())
+                        // e.g. Tree and TreeSet
+                        && rawType.isAssignableFrom(rawType2)
+                        && actualTypeArguments.length == actualTypeArguments2.length) {
+                    // descend into recursion on type arguments
+                    for (int i = 0; i < actualTypeArguments.length; i++) {
+                        return isCompatibleTypes(actualTypeArguments[i], actualTypeArguments2[i]);
+                    }
+                } else {
+                    // owner type, raw type or type arguments length don't match
+                    return false;
+                }
+            }
+        } else if (typeToMock instanceof WildcardType) {
+            WildcardType wildcardTypeToMock = (WildcardType) typeToMock;
+            Type[] upperBounds = wildcardTypeToMock.getUpperBounds();
+            return Arrays.stream(upperBounds).anyMatch(t -> isCompatibleTypes(t, mockType));
+        } else if (typeToMock instanceof GenericArrayType && mockType instanceof GenericArrayType) {
```

Comment:
CodeCov reports this branch is not covered. Can you please expand the coverage here or remove this check?

Reply:
The `WildcardType` part is covered:
![image](https://user-images.githubusercontent.com/1823661/221698483-d4318d2e-69df-4f9c-8255-375f1d772d0c.png)
But you are right that the `GenericTypeArrayType` part from line 64 isn't. I'll remove it, as Mockito cannot mock Array types (as I found out in a test)

Reply:
Removed not covered code

Reply:
There is still some coverage missing, can you either remove the uncovered branches or add regression tests that cover the logic? https://app.codecov.io/gh/mockito/mockito/pull/2923/blob/src/main/java/org/mockito/internal/configuration/injection/filter/TypeBasedCandidateFilter.java Since this is intricate logic and a hot codepath, I would like to see full coverage here.